config/constants/settings/definition.rb

```diff
@@ -0,0 +1,335 @@
+        if @value.respond_to?(:call)
+          @value.call
+        else
+          @value
+        end
+      end
+    end
+
+    def serialized?
+      %i[array hash].include?(format)
+    end
+
+    def writable?
+      if writable.respond_to?(:call)
+        writable.call
+      else
+        !!writable
+      end
+    end
+
+    def override_value(other_value)
+      if format == :hash
+        value.deep_merge! other_value
+      else
+        self.value = other_value
+      end
+
+      raise ArgumentError, "Value for #{name} must be one of #{allowed.join(', ')} but is #{value}" unless valid?
+
+      self.writable = false
+    end
+
+    def valid?
+      # TODO: it would make sense to also check the type of the value (e.g. boolean).
+      # But as using e.g. 0 for a boolean is quite common, that would break.
+      !allowed ||
+        (format == :array && (value - allowed).empty?) ||
+        allowed.include?(value)
+    end
+
+    def allowed
+      if @allowed.respond_to?(:call)
+        @allowed.call
+      else
+        @allowed
+      end
+    end
+
+    class << self
+      # Adds a setting definition to the set of configured definitions. A definition will define a name and a default value.
+      # However, that value can be overwritten by (lower tops higher):
+      # * a value stored in the database (`settings` table)
+      # * a value in the config/configuration.yml file
+      # * a value provided by an ENV var
+      #
+      # @param [Object] name The name of the definition
+      # @param [Object] value The default value the setting has if not overwritten.
+      # @param [nil] format The format the value is in e.g. symbol, array, hash, string. If a value is present,
+      #  the format is deferred.
+      # @param [TrueClass] writable Whether the value can be set in the UI. In case the value is set via file or ENV var,
+      #  this will be set to false later on and UI elements that refer to the definition will be disabled.
+      # @param [nil] allowed The array of allowed values that can be assigned to the definition.
+      #  Will serve to be validated against. A lambda can be provided returning an array in case
+      #  the array needs to be evaluated dynamically. In case of e.g. boolean format, setting
+      #  an allowed array is not necessary.
+      def add(name,
+              value:,
+              format: nil,
+              writable: true,
+              allowed: nil)
+        return if @by_name.present? && @by_name[name.to_s].present?
+
+        @by_name = nil
+
+        definition = new(name,
+                         format: format,
+                         value: value,
+                         writable: writable,
+                         allowed: allowed)
+
+        override_value(definition)
+
+        all << definition
+      end
+
+      def define(&block)
+        instance_exec(&block)
+      end
+
+      def [](name)
+        by_name ||= all.group_by(&:name).transform_values(&:first)
```

Comment:
`index_by` can be used instead of `group_by().transform_values(&:first)`

Reply:
:+1: done